### PR TITLE
Galloping search for binary_search_util

### PR DESCRIPTION
**Pull request by llogiq**
This is unlikely to improve perf much unless for synthetic benchmarks, but I figure it likely won't hurt either.

### CODE REVIEW

**Comment by llogiq** [comment]
Btw. I have a quickcheck test to test against master, but that probably shouldn't run with every build.

File: src/librustc_data_structures/binary_search_util/mod.rs [hunk 1]
```diff
@@ -15,34 +15,49 @@ where
...
-    // here, but it's not really worth it for small-ish
-    // amounts of data.)
-    let mut start = mid;
-    while start > 0 {
-        if key_fn(&data[start - 1]) == *key {
-            start -= 1;
-        } else {
+    // We get back *some* element with the given key -- so do
+    // a galloping search backwards to find the *first* one.
+    let (mut start, mut step, mut previous, size) = (mid, 1, mid, data.len());
```
Comment (Mark-Simulacrum) [inline src/librustc_data_structures/binary_search_util/mod.rs]: Could we declare these on separate lines? My eyes are having trouble correlating initial values with the variable names :)

**Review by Mark-Simulacrum — COMMENTED** [review]
Algorithm looks correct, though I can't say that I can follow it super closely (would likely need to sketch it out on paper to review thoroughly). Since you've tested with quickcheck though and it seems to match my understanding of what it should look like, this seems fine.

r=me with the declarations split onto multiple lines

**Comment by Mark-Simulacrum** [comment]
@bors r+

Thanks!

**Pull request merged**